### PR TITLE
Add `at-rule-descriptor-no-unknown`

### LINKED ISSUE #8148: Add `at-rule-descriptor-no-unknown`
### What is the problem you're trying to solve?

A dedicated rule to flag unknown descriptors within at-rules.

For example, `margin` is not a known descriptor for `@counter-style`:

```css
@counter-style foo { margin: 0; }
```

### What solution would you like to see?

Suggested blueprint:

- Name: `at-rule-descriptor-no-unknown`
- Description: Disallow unknown descriptors for at-rules.
- Message: `Unexpected unknown descriptor "${descriptor}" for at-rule "${at-rule}"`
- Primary option: `true`
- Secondary options: n/a
- Category: Avoid errors > Unknown
- Included in sharable configs?: Yes, recommended.
- Expanded description:

This rule considers values for properties defined within the CSS specifications to be known.

Certain at-rules only allow specific descriptors. This rule checks that. For example, only the `syntax`, `inherits` and 
`initial-value` descriptors are known to the `@property` at-rule.

### PR DESCRIPTION
**Pull request by ryo-manba**
> Which issue, if any, is this issue related to?

Closes #8148 

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/at-rule-descriptor-no-unknown/index.mjs
Comment (ryo-manba) [inline lib/rules/at-rule-descriptor-no-unknown/index.mjs:1]: It looks no type definitions have been added.
See: 
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7f6d332bcd59a641e035ac645259603fa6ed0f01/types/css-tree/index.d.ts

I will submit a PR to upstream.

File: lib/rules/at-rule-descriptor-no-unknown/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/index.mjs:37]:
```suggestion
				if (error.name !== 'SyntaxReferenceError') return;
```

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:104]:
```suggestion
```

These are properties within style rules.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:28]:
```suggestion
		code: '@media (width > 400px) { a { color: red; } }',
```

This should pass as `color` is a property, not a descriptor.

[Test conventions](https://stylelint.io/developer-guide/rules#write-tests).

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:19]:
```suggestion
		// TODO: The following test can be supported once mdn/data, is updated and the css-tree is updated.
		// see: https://github.com/mdn/data/blob/6b4cae7c0aa96c199a982f2b6efeda7ce67eb515/css/at-rules.json
		{
		  code: '@view-transition { navigation: auto; }',
		  skip: true
		},
```

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:10]:
```suggestion
			code: '@counter-style foo { system: cyclic; symbols: url(bar.svg); suffix: " "; }',
```

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:8]: Let's start with a simple property test:

```
a { unknown: 0; }
```

This should be accepted as the rule doesn't check properties.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:29]: Let's add a valid test at the end for a nested media query where a declaration is the child of the at-rule:

```css
a { @media (width > 400px) { color: red; } }
```

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:47]: Let's add a test for multiple warnings:

```css
@property --foo { bar: 0; baz: 0; }
```

And another for a descriptor that's known for one type of at-rule but not another:

```css
@property --foo { system: cyclic; }
```

File: lib/rules/at-rule-descriptor-no-unknown/README.md
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/README.md:16]:
```suggestion
The [`message` secondary option](../../../docs/user-guide/configure.md#message) can accept the arguments of this rule.

This rule checks descriptors within at-rules. To check properties, you can use the [property-no-unknown](../property-no-unknown/README.md) rule.
```

In line with https://github.com/stylelint/stylelint/issues/8150#issuecomment-2536062548

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@ryo-manba, thank you for starting this rule! It's looking great so far.

I've requested a few changes, mostly as suggestions you can commit.

File: docs/user-guide/rules.md
Comment (ryo-manba) [reply docs/user-guide/rules.md:128]: Sorry, I made a mistake.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:98]: Sorry, I still don’t understand. I thought this was a test for `@container` [descriptors](https://drafts.csswg.org/css-conditional-5/#container-descriptor-table). Is this actually testing properties inside `@container` style rules instead?

**Comment by ryo-manba** [comment]
@jeddy3 
Thanks for your kind review!
I have applied your suggestions.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [reply lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:135]: > I thought this was a test for @container [descriptors](https://drafts.csswg.org/css-conditional-5/#container-descriptor-table).

That's interesting. It's the first time I've seen "descriptors" used to describe these. Elsewhere in the spec, they are called [container features](https://drafts.csswg.org/css-conditional-5/#container-features).

Shall we limit this rule to checking if "[descriptor declarations](https://drafts.csswg.org/css-syntax/#declaration)" are known for an at-rule? If so, we can delete these tests.

**Comment by jeddy3** [comment]
@ryo-manba Thank you for making the requested changes. 

I've [answered](https://github.com/stylelint/stylelint/pull/8197#discussion_r1885750817) your question inline.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:135]: Ah, I see! The "descriptors" for `@container` are things like the `width` in this:

```css
@container (width > 600px) {}
```

I misunderstood. Thanks for the detail.

File: lib/rules/at-rule-descriptor-no-unknown/README.md
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/README.md:38]:
```suggestion

<!-- prettier-ignore -->
```

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:24]:
```suggestion
		{
			code: '@view-transition { navigation: auto; }',
		},
		{
			code: '@foo { bar: 0 }',
		},
```

We should ignore these as an unknown (to CSSTree).

File: lib/rules/at-rule-descriptor-no-unknown/index.cjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/index.cjs:39]:
```suggestion

				if (!error.message.startsWith('Unknown at-rule descriptor')) return;
				
```

Currently, the rule also checks and flags unknown at-rules. Let's return early for anything that isn't an unknown descriptor error. This change should make the suggestion to the tests pass.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@ryo-manba Thank you for making the changes.

I've given the code another run through and made a couple more suggestions.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:24]: Ah, I see, that does seem more natural. Thank you for the suggestion!

Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs]: Fixed in this commit.
[Skip unknown at-rules](https://github.com/stylelint/stylelint/pull/8197/commits/d44e66cbe99f7c57e663d196e4642c9cdc264ab3)

File: lib/rules/at-rule-descriptor-no-unknown/index.cjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/index.cjs:39]: Fixed in this commit.
[Skip unknown at-rules](https://github.com/stylelint/stylelint/pull/8197/commits/d44e66cbe99f7c57e663d196e4642c9cdc264ab3)

**Review by jeddy3 — COMMENTED** [review]
LGTM, thank you. 

Let's wait for another review as it's good to get at least another pair of eyes on a pull request that adds a new rule.

File: lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/__tests__/index.mjs:78]:
```suggestion
		},
		{
			code: '@position-try --foo { @supports (bar: 0) { bar: 0 } }',
			message: messages.rejected('@position-try', 'bar'),
			line: 1,
			column: 44,
			endLine: 1,
			endColumn: 50,
			description: 'an unknown descriptor in a nested at-rule',
		},
```

Let's add a test for this, caught in https://github.com/stylelint/stylelint/pull/8203#discussion_r1886867928

**Review by jeddy3 — COMMENTED** [review]
I've suggested another test.

Also, if you've time, can you see if [there's a performance benefit](https://stylelint.io/developer-guide/rules#improve-the-performance-of-a-rule) from not walking nesting supported at-rules, e.g.

```css
root.walkAtRules(/^((?!supports).)*$/i, (atRule) => {
```

Except, construct the regex from our reference list:

https://github.com/stylelint/stylelint/blob/851debebe37bd9fe4de7495430e2b8b1da369710/lib/reference/atKeywords.mjs#L4-L11

Like we do utilities like [`hasColorFunction`](https://github.com/stylelint/stylelint/blob/main/lib/utils/hasColorFunction.mjs).

**Comment by ryo-manba** [comment]
@jeddy3 
Thank you for the suggestion! I ran the benchmarks three times each, and it turns out that skipping nested at-rules improves performance.

### Before

at-rule-descriptor-no-unknown_benchmark_before-skip

### After

at-rule-descriptor-no-unknown_benchmark_after-skip

I’ve added a [commit](https://github.com/stylelint/stylelint/pull/8197/commits/39c023eb5a74aedd5616af0af6d6973a0e11551c) for this change. Please let me know if you have any concerns or further suggestions!

File: lib/rules/at-rule-descriptor-no-unknown/index.mjs
Comment (ybiquitous) [reply lib/rules/at-rule-descriptor-no-unknown/index.mjs:1]: If the upstream PR may be a blocker, we can also work around the type error via patch-package, such as <https://github.com/stylelint/stylelint/pull/7543/files#diff-9005bb639a6220616aebaf301054701ddc479adaac5c8885c0825e735794d234>. Once the upstream PR is merged, we can remove the patch and bump the type package safely. 👌🏼

**Review by ybiquitous — COMMENTED** [review]
Thanks for the pull request! 👍🏼 

I think this is almost good, but let me suggest a few minor improvements.

File: lib/rules/at-rule-descriptor-no-unknown/index.mjs
Comment (jeddy3) [reply lib/rules/at-rule-descriptor-no-unknown/index.mjs:50]: Good catch! Let's make this change.

Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/index.mjs:50]: It was an implementation mistake. Thanks!

File: lib/rules/at-rule-descriptor-no-unknown/index.mjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/index.mjs:1]: Thanks for sharing the steps!
I added in [this commit](https://github.com/stylelint/stylelint/pull/8197/commits/ae422f2fd1a84f75b271f7b84b8344bd83f94c36).

**Review by jeddy3 — APPROVED** [review]
LGTM, thank you!

File: lib/rules/at-rule-descriptor-no-unknown/index.cjs
Comment (jeddy3) [inline lib/rules/at-rule-descriptor-no-unknown/index.cjs:38]:
```suggestion
		root.walkAtRules(UNSUPPORTED_NESTING_REGEX, (atRule) => {
		    if (!isStandardSyntaxAtRule(atRule)) return;
			atRule.walkDecls((decl) => {
			    if (!isStandardSyntaxDeclaration(decl)) return;
				const { prop, value } = decl;
```

Let's return early for non-standard at-rules and decls.

See https://github.com/stylelint/stylelint/pull/8231

**Review by jeddy3 — COMMENTED** [review]
@ryo-manba Sorry for the drip of review comments. It's been a while since I reviewed a new rule, and I'm remembering bits as I review other PRs.

**Comment by jeddy3** [comment]
@ryo-manba You should've received an invite to the Stylelint organisation. Thank you so much for your contributions these past few months; they've been fantastic! As a member, you can now merge your own pull requests, including this one. We typically wait for one approval for simple fixes and two for bigger changes like adding new rules. 

If you want, you can now also help manage [pull requests](https://stylelint.io/maintainer-guide/pull-requests) and [issues](https://stylelint.io/maintainer-guide/issues) across the repos, including assigning labels, changing titles, and using saved replies. You're not obligated to do this, though. As open-source contributors, we do what we can when we can. Welcome to the org!

**Comment by ryo-manba** [comment]
@jeddy3 
Thank you so much! I'm thrilled to become a member!
I'll continue contributing and doing my best to support the community! 😊

**Comment by Mouvedia** [comment]
@ryo-manba https://github.com/stylelint/stylelint/pull/8235#discussion_r1898551642
I was thinking of you when I asked that question.
@jeddy3 beats me to it.

File: lib/rules/at-rule-descriptor-no-unknown/index.cjs
Comment (ryo-manba) [reply lib/rules/at-rule-descriptor-no-unknown/index.cjs]: Applied your suggestion!
[Return early for non-standard at-rules and decls](https://github.com/stylelint/stylelint/pull/8197/commits/f3bc4afae0d0c9d5bd9b81a38ebe037f583e4564)

**Review by ybiquitous — APPROVED** [review]
Sorry for the late approval. It looks great to me. 👍🏼 

@ryo-manba Welcome to the Stylelint team! 😄

**Review by jeddy3 — APPROVED** [review]
Thank you for adding the `isStandardSyntax*` checks.

LGTM. Feel free to squash and merge.

**Pull request merged**
